Gems/Prefab/PrefabDependencyViewer/Code/Source/PrefabDependencyViewerEditorSystemComponent.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Serialization/SerializeContext.h>
+#include <AzToolsFramework/API/ViewPaneOptions.h>
+#include <PrefabDependencyViewerEditorSystemComponent.h>
+#include <MainWindow.h>
+#include <QMenu>
+#include <LyViewPaneNames.h>
+#include <PrefabDependencyTree.h>
+
+namespace PrefabDependencyViewer
+{
+    using Outcome = AZ::Outcome<PrefabDependencyTree, const char*>;
+
+    InstanceEntityMapperInterface* PrefabDependencyViewerEditorSystemComponent::s_prefabEntityMapperInterface = nullptr;
+    PrefabSystemComponentInterface* PrefabDependencyViewerEditorSystemComponent::s_prefabSystemComponentInterface = nullptr;
+    PrefabPublicInterface* PrefabDependencyViewerEditorSystemComponent::s_prefabPublicInterface = nullptr;
+
+    const char* PrefabDependencyViewerEditorSystemComponent::s_prefabViewerTitle = "Prefab Dependencies Viewer";
+
+    void PrefabDependencyViewerEditorSystemComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (auto serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serializeContext->Class<PrefabDependencyViewerEditorSystemComponent, AZ::Component>()
+                ->Version(0);
+        }
+    }
+
+    PrefabDependencyViewerEditorSystemComponent::PrefabDependencyViewerEditorSystemComponent() = default;
+
+    void PrefabDependencyViewerEditorSystemComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided)
+    {
+        provided.push_back(AZ_CRC_CE("PrefabDependencyViewerEditorService"));
+    }
+
+    void PrefabDependencyViewerEditorSystemComponent::GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& incompatible)
+    {
+        incompatible.push_back(AZ_CRC_CE("PrefabDependencyViewerEditorService"));
+    }
+
+    void PrefabDependencyViewerEditorSystemComponent::GetRequiredServices([[maybe_unused]] AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+    }
+
+    void PrefabDependencyViewerEditorSystemComponent::GetDependentServices([[maybe_unused]] AZ::ComponentDescriptor::DependencyArrayType& dependent)
+    {
+    }
+
+    void PrefabDependencyViewerEditorSystemComponent::Activate()
+    {
+        s_prefabEntityMapperInterface = AZ::Interface<InstanceEntityMapperInterface>::Get();
```

Comment:
It's unclear why you would cache the result from `AZ::Interface` like this. What's the reason for taking this approach?

Reply:
Instance entity mapper interface is needed every time the user clicks on View Dependencies. I could retrieve the interface every time this event happens. 

The only drawback is you would make the same function call over and over again. Caching it upfront gets rid of that overhead. Also, singletons are available throughout the entire lifetime of the program so they can't becoming dangling unless something happens to the interface itself.

Reply:
Resolved